File: app/src/main/java/com/example/uscatterbrain/network/BlockSequencePacket.java
package com.example.uscatterbrain.network;

import com.example.uscatterbrain.ScatterProto;
import com.github.davidmoten.rx2.Bytes;
import com.google.protobuf.ByteString;
import com.google.protobuf.GeneratedMessageLite;
import com.goterl.lazycode.lazysodium.interfaces.GenericHash;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.UUID;

import io.reactivex.Completable;
import io.reactivex.Flowable;
import io.reactivex.Observable;
import io.reactivex.Single;

/**
 * Wrapper class for protocol buffer BlockSequence message
 */
public class BlockSequencePacket implements ScatterSerializable {

    private final int mSequenceNumber;
    private final ByteString mData;
    private File mDataOnDisk;
    private final ScatterProto.BlockSequence mBlockSequence;
    private boolean dataNative;
    private UUID luidtag;

    /**
     * Verify the hash of this message against its header
     *
     * @param bd the bd
     * @return boolean whether verification succeeded
     */
    public boolean verifyHash(BlockHeaderPacket bd) {
        byte[] seqnum = ByteBuffer.allocate(4).putInt(this.mSequenceNumber).order(ByteOrder.BIG_ENDIAN).array();

        byte[] data;
        if (this.mBlockSequence.getDataCase() == ScatterProto.BlockSequence.DataCase.DATA_CONTENTS ) {
            data = this.mBlockSequence.getDataContents().toByteArray();
            this.dataNative = false;
        } else {
            this.dataNative = true;
            data = new byte[0];
        }
        byte[] testhash = new byte[GenericHash.BYTES];
        byte[] state = new byte[LibsodiumInterface.getSodium().crypto_generichash_statebytes()];
        LibsodiumInterface.getSodium().crypto_generichash_init(state,null, 0, testhash.length);
        LibsodiumInterface.getSodium().crypto_generichash_update(state, seqnum, seqnum.length);
        LibsodiumInterface.getSodium().crypto_generichash_update(state, data, data.length);
        LibsodiumInterface.getSodium().crypto_generichash_final(state, testhash, testhash.length);
        return LibsodiumInterface.getSodium().sodium_compare(testhash, bd.getHash(this.mSequenceNumber).toByteArray(), testhash.length) == 0;
    }

    /**
     * Calculates the hash of this message
     *
     * @return the hash
     */
    public byte[] calculateHash() {
        byte[] hashbytes = new byte[GenericHash.BYTES];
        byte[] state = new byte[LibsodiumInterface.getSodium().crypto_generichash_statebytes()];
        byte[] seqnum = ByteBuffer.allocate(4).putInt(this.mSequenceNumber).order(ByteOrder.BIG_ENDIAN).array();
        LibsodiumInterface.getSodium().crypto_generichash_init(state, null, 0, hashbytes.length);
        LibsodiumInterface.getSodium().crypto_generichash_update(state, seqnum, seqnum.length);
        LibsodiumInterface.getSodium().crypto_generichash_update(state, mData.toByteArray(), mData.size());
        LibsodiumInterface.getSodium().crypto_generichash_final(state, hashbytes, hashbytes.length);
        return hashbytes;
    }

    public boolean isNative() {
        return dataNative;
    }

    /**
     * Calculates the hash of this message
     *
     * @return hash as ByteString
     */
    public ByteString calculateHashByteString() {
        return ByteString.copyFrom(calculateHash());
    }

    @Override
    public GeneratedMessageLite getMessage() {
        return mBlockSequence;
    }

    @Override
    public byte[] getBytes() {
        ByteArrayOutputStream os = new ByteArrayOutputStream();
        try {
            CRCProtobuf.writeToCRC(mBlockSequence, os);
            return os.toByteArray();
        } catch (IOException e) {
            return null;
        }
    }

    @Override
    public ByteString getByteString() {
        return ByteString.copyFrom(getBytes());
    }

    @Override
    public Completable writeToStream(OutputStream os) {
        return Completable.fromAction(() -> CRCProtobuf.writeToCRC(mBlockSequence, os));
    }

    @Override
    public Flowable<byte[]> writeToStream(int fragsize) {
        return Bytes.from(new ByteArrayInputStream(getBytes()), fragsize);
    }

    @Override
    public PacketType getType() {
        return PacketType.TYPE_BLOCKSEQUENCE;
    }

    @Override
    public void tagLuid(UUID luid) {
        luidtag = luid;
    }

    @Override
    public UUID getLuid() {
        return luidtag;
    }

    private BlockSequencePacket(InputStream is) throws IOException {
        this.mBlockSequence = CRCProtobuf.parseFromCRC(ScatterProto.BlockSequence.parser(), is);
        if (mBlockSequence.getDataCase() == ScatterProto.BlockSequence.DataCase.DATA_CONTENTS) {
            this.mData = mBlockSequence.getDataContents();
            this.dataNative = false;
        } else {
            this.mData = ByteString.EMPTY;
            this.dataNative = true;
        }
        this.mSequenceNumber = mBlockSequence.getSeqnum();
    }

    /**
     * Parse from block sequence packet.
     *
     * @param is the is
     * @return the block sequence packet
     */
    public static Single<BlockSequencePacket> parseFrom(InputStream is) {
        return Single.fromCallable(() -> new BlockSequencePacket(is));
    }

     public static Single<BlockSequencePacket> parseFrom(Observable<byte[]> flowable) {
        InputStreamObserver observer = new InputStreamObserver();
        flowable.subscribe(observer);
        return BlockSequencePacket.parseFrom(observer).doFinally(observer::close);
     }

    public static Single<BlockSequencePacket> parseFrom(Flowable<byte[]> flowable) {
        InputStreamFlowableSubscriber observer = new InputStreamFlowableSubscriber();
        flowable.subscribe(observer);
        return BlockSequencePacket.parseFrom(observer).doFinally(observer::close);
    }

    private BlockSequencePacket(Builder builder) {
        this.mSequenceNumber = builder.getmSequenceNumber();
        ByteString d = builder.getmData();
        this.mDataOnDisk = builder.getmDataOnDisk();
        ScatterProto.BlockSequence.Builder tmpbuilder = ScatterProto.BlockSequence.newBuilder();
        if (d != null) {
            tmpbuilder.setDataContents(d);
            this.mData = d;
            this.dataNative = false;
        } else {
            this.dataNative = true;
            this.mData = ByteString.EMPTY;
        }
        this.mBlockSequence = tmpbuilder.setSeqnum(this.mSequenceNumber)
                .build();


    }

    /**
     * Gets sequence number.
     *
     * @return the sequence number
     */
    public int getmSequenceNumber() {
        return mSequenceNumber;
    }

    /**
     * Gets data.
     *
     * @return the data
     */
    public ByteString getmData() {
        return mData;
    }

    /**
     * Gets data on disk.
     *
     * @return the data on disk
     */
    public File getmDataOnDisk() {
        return mDataOnDisk;
    }

    /**
     * Gets block sequence.
     *
     * @return the block sequence
     */
    public ScatterProto.BlockSequence getmBlockSequence() {
        return mBlockSequence;
    }

    /**
     * New builder builder.
     *
     * @return the builder
     */
    public static Builder newBuilder() {
        return new Builder();
    }

    /**
     * Builder class for BlockSequencePacket
     */
    public static class Builder {

        private int mSequenceNumber;
        private ByteString mData;
        private File mDataOnDisk;
        private boolean mOnDisk;

        /**
         * Instantiates a new Builder.
         */
        public Builder() {

        }

        /**
         * Sets sequence number.
         *
         * @param sequenceNumber the sequence number
         * @return the sequence number
         */
        public Builder setSequenceNumber(int sequenceNumber) {
            this.mSequenceNumber = sequenceNumber;
            return this;
        }

        /**
         * Sets data.
         *
         * @param data the data
         * @return the data
         */
        public Builder setData(ByteString data) {
            this.mData = data;
            return this;
        }

        /**
         * Build block sequence packet.
         *
         * @return the block sequence packet
         */
        public BlockSequencePacket build() {
            return new BlockSequencePacket(this);
        }

        /**
         * Gets sequence number.
         *
         * @return the sequence number
         */
        public int getmSequenceNumber() {
            return mSequenceNumber;
        }

        /**
         * Gets data.
         *
         * @return the data
         */
        public ByteString getmData() {
            return mData;
        }

        /**
         * Gets file to write packet to
         *
         * @return file object
         */
        public File getmDataOnDisk() {
            return mDataOnDisk;
        }}

}

File: app/src/main/java/net/ballmerlabs/uscatterbrain/network/BlockSequencePacket.kt
package net.ballmerlabs.uscatterbrain.network

import com.github.davidmoten.rx2.Bytes
import com.google.protobuf.ByteString
import com.goterl.lazycode.lazysodium.interfaces.GenericHash
import io.reactivex.Completable
import io.reactivex.Flowable
import io.reactivex.Observable
import io.reactivex.Single
import net.ballmerlabs.uscatterbrain.ScatterProto.BlockSequence
import net.ballmerlabs.uscatterbrain.network.ScatterSerializable.PacketType
import java.io.*
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.*

/**
 * Wrapper class for protocol buffer BlockSequence message
 */
class BlockSequencePacket private constructor(builder: Builder) : ScatterSerializable {
    private val mSequenceNumber: Int
    private val mData: ByteString?
    private val mDataOnDisk: File?
    private val mBlockSequence: BlockSequence?
    var isNative = false
        private set
    override var luid: UUID? = null
        private set


    init {
        mSequenceNumber = builder.getmSequenceNumber()
        val d = builder.getmData()
        mDataOnDisk = builder.getmDataOnDisk()
        val tmpbuilder = BlockSequence.newBuilder()
        if (d != null) {
            tmpbuilder.dataContents = d
            mData = d
            isNative = false
        } else {
            isNative = true
            mData = ByteString.EMPTY
        }
        mBlockSequence = tmpbuilder.setSeqnum(mSequenceNumber)
                .build()
    }

    /**
     * Verify the hash of this message against its header
     *
     * @param bd the bd
     * @return boolean whether verification succeeded
     */
    fun verifyHash(bd: BlockHeaderPacket?): Boolean {
        val seqnum = ByteBuffer.allocate(4).putInt(mSequenceNumber).order(ByteOrder.BIG_ENDIAN).array()
        val data: ByteArray
        if (mBlockSequence!!.dataCase == BlockSequence.DataCase.DATA_CONTENTS) {
            data = mBlockSequence.dataContents.toByteArray()
            isNative = false
        } else {
            isNative = true
            data = ByteArray(0)
        }
        val testhash = ByteArray(GenericHash.BYTES)
        val state = ByteArray(LibsodiumInterface.sodium.crypto_generichash_statebytes())
        LibsodiumInterface.sodium.crypto_generichash_init(state, null, 0, testhash.size)
        LibsodiumInterface.sodium.crypto_generichash_update(state, seqnum, seqnum.size.toLong())
        LibsodiumInterface.sodium.crypto_generichash_update(state, data, data.size.toLong())
        LibsodiumInterface.sodium.crypto_generichash_final(state, testhash, testhash.size)
        return LibsodiumInterface.sodium.sodium_compare(testhash, bd!!.getHash(mSequenceNumber).toByteArray(), testhash.size) == 0
    }

    /**
     * Calculates the hash of this message
     *
     * @return the hash
     */
    fun calculateHash(): ByteArray {
        val hashbytes = ByteArray(GenericHash.BYTES)
        val state = ByteArray(LibsodiumInterface.sodium.crypto_generichash_statebytes())
        val seqnum = ByteBuffer.allocate(4).putInt(mSequenceNumber).order(ByteOrder.BIG_ENDIAN).array()
        LibsodiumInterface.sodium.crypto_generichash_init(state, null, 0, hashbytes.size)
        LibsodiumInterface.sodium.crypto_generichash_update(state, seqnum, seqnum.size.toLong())
        LibsodiumInterface.sodium.crypto_generichash_update(state, mData!!.toByteArray(), mData!!.size().toLong())
        LibsodiumInterface.sodium.crypto_generichash_final(state, hashbytes, hashbytes.size)
        return hashbytes
    }

    /**
     * Calculates the hash of this message
     *
     * @return hash as ByteString
     */
    fun calculateHashByteString(): ByteString {
        return ByteString.copyFrom(calculateHash())
    }

    override val bytes: ByteArray
        get() {
            val os = ByteArrayOutputStream()
            return try {
                CRCProtobuf.writeToCRC(mBlockSequence, os)
                os.toByteArray()
            } catch (e: IOException) {
                byteArrayOf(0) //this should be unreachable
            }
        }

    override val byteString: ByteString
        get() = ByteString.copyFrom(bytes)

    override fun writeToStream(os: OutputStream): Completable {
        return Completable.fromAction { CRCProtobuf.writeToCRC(mBlockSequence, os) }
    }

    override fun writeToStream(fragsize: Int): Flowable<ByteArray> {
        return Bytes.from(ByteArrayInputStream(bytes), fragsize)
    }

    override val type: PacketType
        get() = PacketType.TYPE_BLOCKSEQUENCE

    override fun tagLuid(luid: UUID?) {
        this.luid = luid
    }

    /**
     * Gets sequence number.
     *
     * @return the sequence number
     */
    fun getmSequenceNumber(): Int {
        return mSequenceNumber
    }

    /**
     * Gets data.
     *
     * @return the data
     */
    fun getmData(): ByteString? {
        return mData
    }

    /**
     * Gets data on disk.
     *
     * @return the data on disk
     */
    fun getmDataOnDisk(): File? {
        return mDataOnDisk
    }

    /**
     * Gets block sequence.
     *
     * @return the block sequence
     */
    fun getmBlockSequence(): BlockSequence? {
        return mBlockSequence
    }

    /**
     * Builder class for BlockSequencePacket
     */
    class Builder
    /**
     * Instantiates a new Builder.
     */
    {
        private var mSequenceNumber = 0
        private var mData: ByteString? = null
        private val mDataOnDisk: File? = null
        private val mOnDisk = false

        /**
         * Sets sequence number.
         *
         * @param sequenceNumber the sequence number
         * @return the sequence number
         */
        fun setSequenceNumber(sequenceNumber: Int): Builder {
            mSequenceNumber = sequenceNumber
            return this
        }

        /**
         * Sets data.
         *
         * @param data the data
         * @return the data
         */
        fun setData(data: ByteString?): Builder {
            mData = data
            return this
        }

        /**
         * Build block sequence packet.
         *
         * @return the block sequence packet
         */
        fun build(): BlockSequencePacket {
            return BlockSequencePacket(this)
        }

        /**
         * Gets sequence number.
         *
         * @return the sequence number
         */
        fun getmSequenceNumber(): Int {
            return mSequenceNumber
        }

        /**
         * Gets data.
         *
         * @return the data
         */
        fun getmData(): ByteString? {
            return mData
        }

        /**
         * Gets file to write packet to
         *
         * @return file object
         */
        fun getmDataOnDisk(): File? {
            return mDataOnDisk
        }
    }

    companion object {

        private fun builderFromIs(inputStream: InputStream) : Builder {
            val blocksequence = CRCProtobuf.parseFromCRC(BlockSequence.parser(), inputStream)
            val builder = Builder()
            if (blocksequence.dataCase == BlockSequence.DataCase.DATA_CONTENTS) {
                builder.setData(blocksequence.dataContents)
            } else {
                builder.setData(null)
            }
            builder.setSequenceNumber(blocksequence.seqnum)
            return builder
        }

        /**
         * Parse from block sequence packet.
         *
         * @param is the is
         * @return the block sequence packet
         */
        fun parseFrom(`is`: InputStream): Single<BlockSequencePacket> {
            return Single.fromCallable { BlockSequencePacket(builderFromIs(`is`)) }
        }

        fun parseFrom(flowable: Observable<ByteArray>): Single<BlockSequencePacket> {
            val observer = InputStreamObserver(32768) //TODO: find a way to calculate max size
            flowable.subscribe(observer)
            return parseFrom(observer).doFinally { observer.close() }
        }

        fun parseFrom(flowable: Flowable<ByteArray>): Single<BlockSequencePacket> {
            val observer = InputStreamFlowableSubscriber(32768) //TODO: find a way to calculate max size
            flowable.subscribe(observer)
            return parseFrom(observer).doFinally { observer.close() }
        }

        /**
         * New builder builder.
         *
         * @return the builder
         */
        fun newBuilder(): Builder {
            return Builder()
        }
    }
}